Add freezing capability to UTXO

MichaelKim20 opened the pull request:
Add freezing capability to UTXO #204 of part

### Transaction & UTXO in Freeze
The design was based on **Definition of Done**.
Illustrated the changing situation over time for **Block / Transaction / UTXOSet.**
<img width="920" alt="Transaction & UTXO in Freeze" src="https://user-images.githubusercontent.com/11639939/65112733-9c22a880-da1b-11e9-9f1b-55f4b00098c5.png">

Implement validation when receive transaction 'freeze' in melted and , receive transaction 'payment' in frozen.(TX2, TX3 in upper diagram)
Implement validation when receive transaction 'payment' in various status (melting, melted)(TX4, TX5 in upper diagram)

MichaelKim20 commented:
Ready for review

MichaelKim20 commented:
<img width="896" alt="" src="https://user-images.githubusercontent.com/11639939/65110791-82319780-da14-11e9-9367-2f8eac1c87f0.png">

Geod24 commented on source/agora/consensus/data/UTXOSet.d:
```diff
@@ -148,23 +148,108 @@ public class UTXOSet
+        if (tx.type == TxType.Freeze)
         {
+            if ((tx.inputs.length != 1) || (tx.outputs.length != 1))
+                assert(0);
```
Why the limitation on number of inputs and outputs ?

MichaelKim20 replied:
There is the following in Definition of Done
`The amount must be exactly 40,000 BOA (for the moment)`
I thought of simple input and output.
Should I set the sum of input and output to 40,000 BOA?

Geod24 replied:
https://github.com/bpfkorea/agora/blob/3079841a13ab18ebcf7997cdbf682acd9fde730f/source/agora/common/Amount.d#L37-L38

MichaelKim20 replied:
Restricting the number of inputs and outputs is a limitation because the various types of input and output cannot be put in.

Geod24 commented on source/agora/consensus/data/UTXOSet.d:
```diff
@@ -148,23 +148,108 @@ public class UTXOSet
+
+            UTXOSetValue prev_utxo;
+            if (!this.getUTXOSetValue(input.previous, input.index, prev_utxo))
+                assert(0);
```
We can assert on internal failure, but we should never assert on external parameter.
If someone was to send us a malformed transaction, a node would crash, which is not acceptable.

MichaelKim20 replied:
I already knew.
After completing the transaction validation (isInvalidReason), the code was deleted at (Comit SHA: b4c2dff751366ba4a3fa83ceae12bfc3c9390fde)

Geod24 commented on source/agora/consensus/Validation.d:
```diff
@@ -39,8 +39,9 @@ public alias UTXOFinder = scope bool delegate (Hash hash, size_t index,
-    @safe nothrow
+public string isInvalidReason (const Transaction tx,
+                                UTXOFinder findUTXO,
+                                const ulong height = 0) @safe nothrow
```
We shouldn't use default values here

MichaelKim20 replied:
I did so to minimize the code change.
I'll fix it.

Geod24 commented on source/agora/consensus/Validation.d:
```diff
@@ -57,18 +57,72 @@ public string isInvalidReason (const Transaction tx,
 
-        if (!sum_unspent.add(value.output.value))
-            return "Transaction: Input overflow";
+            if (utxo_value.output.value != Amount(40_000))
```
The unit is 10^7 so you're missing a few 0

MichaelKim20 replied:
Oh, I see. I have to fix it.

MichaelKim20 replied:
I think we need to increase the amount of Genesis Block Coins.

Geod24 reviewed (commented):
Wow okay this is a lot of code, we're going to have to break it down a bit.
I see some commit are doing back-and-forth, e.g. a duplicated method is added in one commit and removed 2 commits later, which is a red flag.
Additionally, some functions are doing the following:
```D
void myFunc ()
{
   if (A)
   {

    }
    else
    {

     }
}
```

This essentially means you have 2 functions in one.

Geod24 commented on source/agora/consensus/data/UTXOSet.d:
```diff
@@ -27,6 +27,73 @@ import std.file;
+
+    ***************************************************************************/
+
+    public void computeHash (scope HashDg dg) const nothrow @safe @nogc
```
Why do we need the hash of this struct ?

MichaelKim20 commented:
> Wow okay this is a lot of code, we're going to have to break it down a bit.
> I see some commit are doing back-and-forth, e.g. a duplicated method is added in one commit and removed 2 commits later, which is a red flag.
> Additionally, some functions are doing the following:
> 
> ```d
> void myFunc ()
> {
>    if (A)
>    {
> 
>     }
>     else
>     {
> 
>      }
> }
> ```
> 
> This essentially means you have 2 functions in one.



This may be due to the addition of a function to 'isInvalidReason' to remove redundant functions from 'UTXOSet.updateUTXOCache'.
It was to speed up the process.
But now it will be difficult to confirm, as the whole changed source is committed.

Geod24 commented on source/agora/consensus/data/UTXOSet.d:
```diff
@@ -27,6 +27,56 @@ import std.file;
+    TxType type;
+
+    /// Previous transaction type
+    TxType prev_type;
```
What is this field for ?

MichaelKim20 replied:
Type of previous transaction.
To verify that the freeze is a UTXO just melted.

MichaelKim20 replied:
Please refer to the table below.
<img width="896" alt="Freezing status" src="https://user-images.githubusercontent.com/11639939/64929810-bd419880-d865-11e9-9165-e6c068a7527e.png">

MichaelKim20 commented:
@Geod24 Ready for review

Geod24 commented:
@MukeunKim : I think the first 3 commits can go in their own PR and be merged already, if you want.

Geod24 commented:
I'm not so sure about the last two commits. Do they belong to this PR ?

MichaelKim20 commented:
The first three commit are already included in PR #337.
Please check there first.

MichaelKim20 commented:
![Block_Tx_UTXO 001](https://user-images.githubusercontent.com/11639939/65367182-a7b6df00-dc68-11e9-9e56-224fe5c6b5ff.jpeg)

![Block_Tx_UTXO](https://user-images.githubusercontent.com/11639939/65367153-219a9880-dc68-11e9-89ae-39ac288ef2ee.gif)

![FreezingAndUTXO](https://user-images.githubusercontent.com/11639939/65421822-b7d4e700-de3f-11e9-81ae-fd2d30e40e67.gif)

MichaelKim20 commented:
Ready for review ~

MichaelKim20 committed:
Add `block height` to the parameter of the transaction validation

Block height is used to check if it has changed from 'melting' to 'melted'

Co-Authored-By: HyeonyeobKim <hyeonyeob.kim@bpfkorea.org>

MichaelKim20 committed:
Implement validation of transactions associated with freezing

when receive transaction 'freeze' in melted
when receive transaction 'payment' in frozen
when receive transaction 'payment' in melting, melted

Co-Authored-By: HyeonyeobKim <hyeonyeob.kim@bpfkorea.org>

MichaelKim20 committed:
Add test for freezing and melting while creating blocks

Co-Authored-By: HyeonyeobKim <hyeonyeob.kim@bpfkorea.org>

Geod24 reviewed (approved):
LGTM. Any last comment @AndrejMitrovic ?

AndrejMitrovic commented:
> LGTM. Any last comment @AndrejMitrovic ?

Yes, I will review it now.

AndrejMitrovic commented:
I suggest to add this to Amount.d:

```D
struct Amount
{
    ...
    /// Exact amount that needs to be staked to make a freezing transaction
    public static immutable Amount FreezeAmount =
    Amount(UnitPerCoin.value * 40_000, true);
...
```

And also improving the validation function a little bit, to remove code duplication:

```D
public string isInvalidReason (const Transaction tx, UTXOFinder findUTXO,
    const ulong height)
    @safe nothrow
{
    import std.conv;

    if (tx.inputs.length == 0)
        return "Transaction: No input";

    if (tx.outputs.length == 0)
        return "Transaction: No output";

    // disallow negative amounts
    foreach (output; tx.outputs)
        if (!output.value.isValid())
            return "Transaction: Output(s) overflow or underflow";

    const tx_hash = hashFull(tx);

    string isInvalidInput (Input input, ref UTXOSetValue utxo_value,
        ref Amount sum_unspent)
    {
        if (!findUTXO(input.previous, input.index, utxo_value))
            return "Transaction: Input ref not in UTXO";

        if (!utxo_value.output.address.verify(input.signature, tx_hash[]))
            return "Transaction: Input has invalid signature";

        if (!sum_unspent.add(utxo_value.output.value))
            return "Transaction: Input overflow";

        return null;
    }

    Amount sum_unspent;

    if (tx.type == TxType.Freeze)
    {
        foreach (input; tx.inputs)
        {
            UTXOSetValue utxo_value;
            if (auto fail_reason = isInvalidInput(input, utxo_value, sum_unspent))
                return fail_reason;

            if (utxo_value.type != TxType.Payment)
                return "Transaction: Can only freeze a Payment transaction";
        }

        if (sum_unspent != Amount.FreezeAmount)
            return "Transaction: Only available when the amount is 40,000";
    }
    else if (tx.type == TxType.Payment)
    {
        uint count_freeze = 0;
        foreach (input; tx.inputs)
        {
            UTXOSetValue utxo_value;
            if (auto fail_reason = isInvalidInput(input, utxo_value, sum_unspent))
                return fail_reason;

            // when status is frozen, it will begin to melt
            // In this case, all inputs must be frozen.
            if (utxo_value.type == TxType.Freeze)
                count_freeze++;

            // when status is (frozen->melting->melted) or (frozen->melting)
            if (utxo_value.type == TxType.Payment)
            {
                // when status is still melting
                if (height < utxo_value.unlock_height)
                    return "Transaction: Not available when melting UTXO";
            }
        }

        // current limitation: if any UTXO is frozen, they all must be frozen
        if ((count_freeze > 0) && (count_freeze != tx.inputs.length))
            return "Transaction: Rejected combined inputs (freeze & payment)";
    }
    else assert(0, "Unhandled transaction type: " ~ tx.to!string);

    Amount new_unspent;
    if (!tx.getSumOutput(new_unspent))
        return "Transaction: Referenced Output(s) overflow";

    if (!sum_unspent.sub(new_unspent))
        return "Transaction: Output(s) are higher than Input(s)";

    return null;
}
```

Here is the diff:

```diff
diff --git a/source/agora/common/Amount.d b/source/agora/common/Amount.d
index 26912a9b..911dca3c 100644
--- a/source/agora/common/Amount.d
+++ b/source/agora/common/Amount.d
@@ -41,6 +41,9 @@ public struct Amount
     /// Maximum amount of money that can ever be in circulation
     public static immutable Amount MaxUnitSupply =
         Amount(UnitPerCoin.value * MaxCoinSupply, true);
+    /// Exact amount that needs to be staked to make a freezing transaction
+    public static immutable Amount FreezeAmount =
+        Amount(UnitPerCoin.value * 40_000, true);
 
     /// Helper type for `toString`
     private alias SinkT = void delegate(scope const(char)[] v) @safe;
diff --git a/source/agora/consensus/Validation.d b/source/agora/consensus/Validation.d
index 0b35bfd8..fa16ff63 100644
--- a/source/agora/consensus/Validation.d
+++ b/source/agora/consensus/Validation.d
@@ -40,9 +40,12 @@ public alias UTXOFinder = scope bool delegate (Hash hash, size_t index,
 
 *******************************************************************************/
 
-public string isInvalidReason (const Transaction tx, UTXOFinder findUTXO, const ulong height)
+public string isInvalidReason (const Transaction tx, UTXOFinder findUTXO,
+    const ulong height)
     @safe nothrow
 {
+    import std.conv;
+
     if (tx.inputs.length == 0)
         return "Transaction: No input";
 
@@ -54,73 +57,76 @@ public string isInvalidReason (const Transaction tx, UTXOFinder findUTXO, const
         if (!output.value.isValid())
             return "Transaction: Output(s) overflow or underflow";
 
-    Amount sum_unspent;
-
     const tx_hash = hashFull(tx);
 
+    string isInvalidInput (Input input, ref UTXOSetValue utxo_value,
+        ref Amount sum_unspent)
+    {
+        if (!findUTXO(input.previous, input.index, utxo_value))
+            return "Transaction: Input ref not in UTXO";
+
+        if (!utxo_value.output.address.verify(input.signature, tx_hash[]))
+            return "Transaction: Input has invalid signature";
+
+        if (!sum_unspent.add(utxo_value.output.value))
+            return "Transaction: Input overflow";
+
+        return null;
+    }
+
+    Amount sum_unspent;
+
     if (tx.type == TxType.Freeze)
     {
         foreach (input; tx.inputs)
         {
-            // all referenced outputs must be present
             UTXOSetValue utxo_value;
-            if (!findUTXO(input.previous, input.index, utxo_value))
-                return "Transaction: Input ref not in UTXO";
-
-            if (!utxo_value.output.address.verify(input.signature, tx_hash[]))
-                return "Transaction: Input has invalid signature";
-
-            if (!sum_unspent.add(utxo_value.output.value))
-                return "Transaction: Input overflow";
+            if (auto fail_reason = isInvalidInput(input, utxo_value, sum_unspent))
+                return fail_reason;
 
             if (utxo_value.type != TxType.Payment)
-                return "Transaction: Only available freeze on TxType.Payment";
+                return "Transaction: Can only freeze a Payment transaction";
         }
 
-        if (sum_unspent != Amount(400_000_000_000L))
+        if (sum_unspent != Amount.FreezeAmount)
             return "Transaction: Only available when the amount is 40,000";
     }
     else if (tx.type == TxType.Payment)
     {
         uint count_freeze = 0;
-
         foreach (input; tx.inputs)
         {
-            // all referenced outputs must be present
             UTXOSetValue utxo_value;
-            if (!findUTXO(input.previous, input.index, utxo_value))
-                return "Transaction: Input ref not in UTXO";
+            if (auto fail_reason = isInvalidInput(input, utxo_value, sum_unspent))
+                return fail_reason;
 
-            if (!utxo_value.output.address.verify(input.signature, tx_hash[]))
-                return "Transaction: Input has invalid signature";
-
-            if (!sum_unspent.add(utxo_value.output.value))
-                return "Transaction: Input overflow";
-
-            //  when status is frozen, it will begin to melt
-            //  In this case, all inputs must be frozen.
+            // when status is frozen, it will begin to melt
+            // In this case, all inputs must be frozen.
             if (utxo_value.type == TxType.Freeze)
                 count_freeze++;
 
-            //  when status is (frozen->melting->melted) or (frozen->melting)
+            // when status is (frozen->melting->melted) or (frozen->melting)
             if (utxo_value.type == TxType.Payment)
             {
-                //  when status is still melting
+                // when status is still melting
                 if (height < utxo_value.unlock_height)
                     return "Transaction: Not available when melting UTXO";
             }
         }
 
-        //  In this case, all inputs must be frozen.
+        // current limitation: if any UTXO is frozen, they all must be frozen
         if ((count_freeze > 0) && (count_freeze != tx.inputs.length))
-            return "Transaction: TxType.Freeze and Payment are not handled together in Input";
+            return "Transaction: Rejected combined inputs (freeze & payment)";
     }
+    else assert(0, "Unhandled transaction type: " ~ tx.to!string);
 
     Amount new_unspent;
     if (!tx.getSumOutput(new_unspent))
         return "Transaction: Referenced Output(s) overflow";
+
     if (!sum_unspent.sub(new_unspent))
         return "Transaction: Output(s) are higher than Input(s)";
+
     return null;
 }

```

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -576,7 +578,7 @@ unittest
 
         // always use the same amount, for simplicity
-        const Amount AmountPerTx = Amount(40_000);
+        const Amount AmountPerTx = Amount(400_000_000_000L);
```
Once the freeze amount is added to the `Amount` struct, it can be used here directly.

AndrejMitrovic commented:
I will submit a PR that slightly improves the code.
